src/dag_algo/longest_path.rs

```diff
@@ -28,52 +26,22 @@ where
     T: Num + Zero + PartialOrd + Copy,
 {
     let dag = &graph.graph;
-    let mut path: Vec<usize> = Vec::new();
-    let nodes = match algo::toposort(&graph.graph, None) {
-        Ok(nodes) => nodes,
-        Err(_err) => return Err(DAGHasCycle::new_err("Sort encountered a cycle")),
-    };
-    if nodes.is_empty() {
-        return Ok((path, T::zero()));
-    }
-    let mut dist: HashMap<NodeIndex, (T, NodeIndex)> = HashMap::new();
-    for node in nodes {
-        let parents = dag.edges_directed(node, petgraph::Direction::Incoming);
-        let mut us: Vec<(T, NodeIndex)> = Vec::new();
-        for p_edge in parents {
-            let p_node = p_edge.source();
-            let weight: T = weight_fn(p_node.index(), p_edge.target().index(), p_edge.weight())?;
-            let length = dist[&p_node].0 + weight;
-            us.push((length, p_node));
+
+    // Create a new weight function that matches the required signature
+    let edge_cost = |edge_ref: EdgeReference<'_, PyObject>| -> T {
+        let source = edge_ref.source().index();
+        let target = edge_ref.target().index();
+        let weight = edge_ref.weight();
+        match weight_fn(source, target, weight) {
+            Ok(w) => w,
+            Err(_) => T::zero(),
```

Comment:
I don't think this the correct way to handle invalid edge weights, but I am unsure how to fix it. Currently we fail the test: 
```
FAIL: test_nan_not_valid_weight (digraph.test_depth.TestWeightedLongestPath.test_nan_not_valid_weight)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/henryzou/Documents/GitHub/rustworkx/tests/digraph/test_depth.py", line 287, in test_nan_not_valid_weight
    with self.assertRaises(ValueError):
AssertionError: ValueError not raised
```

where in the test the weight function is:

``` python
def weight_fn(*_):
            return float("nan")
 ```